index.src.html

```diff
@@ -750,6 +756,130 @@ <h3 id="try-delivery" algorithm>
 </section>
 
 <section>
+  <h2 id="observers">Reporting Observers</h2>
+  Each <a>environment settings object</a> has an associated list of
+  <dfn>registered reporting observers</dfn> ({{ReportingObserver}} objects).
+   
+  <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
+
+  <pre class="idl">
+[NoInterfaceObject]
+interface ReportBody {
+};
+
+[NoInterfaceObject]
+interface Report {
+    readonly attribute DOMString type;
+    readonly attribute DOMString url;
+    readonly attribute ReportBody? body;
+};
+
+[Constructor(ReportingObserverCallback callback)]
+interface ReportingObserver {
+    void observe(optional ReportingObserverOptions options);
+    void disconnect();
+};
+
+callback ReportingObserverCallback = void (sequence&lt;Report> reports, ReportingObserver observer);
+
+dictionary ReportingObserverOptions {
+  sequence&lt;DOMString> types;
+};
+  </pre>
+  A {{ReportingObserver}} object can be used to observe some types of
+  <a>reports</a> from JavaScript when they are created.
+  
+  Issue: How do we specify which reports are observable?  Network-layer reports
+  like CSP reports will not be.  Maybe there's something about the settings
+  object?
+  
+  A <dfn id=dom-report interface>Report</dfn> is the application exposed
+  representation of a <a>report</a>. <dfn attribute for="Report">type</dfn>
+  returns [=report/type=], <dfn attribute for="Report">url</dfn> returns
+  [=report/url=] and <dfn attribute for="Report">body</dfn> returns
+  [=report/body=].
+    
+  Each {{ReportingObserver}} object has these associated concepts:
+    - A <dfn for=ReportingObserver>callback</dfn> set on creation.
+    - An {{ReportingObserverOptions}} dictionary called
+      <dfn for=ReportingObserver>options</dfn>.
+    - A list of {{Report}} objects called the <dfn for=ReportingObserver>report
+      list</dfn> that is initially empty.
+
+  The <dfn constructor for=ReportingObserver><code>
+  ReportingObserver(|callback|)</code></dfn> constructor, when invoked, must
+  create a new {{ReportingObserver}} object with <a>callback</a> set to
+  |callback|, and then return it.
+  
+  The <dfn method for=ReportingObserver><code>observe(|options|)</code></dfn>
+  method, when invoked, must run these steps:
+  
+  1. Set the <a>context object</a>'s <a>options</a> to |options|.
+  
+  2. If it is not already present, add the <a>context object</a> to the 
+     list of <a>registered reporting observers</a> for the <a>environment
+     settings object</a> associated with the <a>context object</a>.
+  
+  The <dfn method for=ReportingObserver><code>disconnect()</code></dfn> method,
+  when invoked, must run these steps:
+  
+  1. If present, remove the <a>context object</a> from the list of
+     <a>registered reporting observers</a> for the <a>environment settings
+     object</a> associated with the <a>context object</a>.
+ 
```

Comment:
Based on experience with PerformanceObserver (see https://github.com/w3c/performance-timeline/issues/74), we should also support takeRecords().

Reply:
Another thing to consider: what about reports queued during "boot time" of the page, before observer can be registered? E.g. it's best practice to remove analytics and other non-critical work from the critical path, but that means that I may miss a whole lot of reports. 

For PerfObserver, we're adding support for buffering events up to a certain point: https://github.com/w3c/performance-timeline/issues/81

Reply:
Thanks!
Discussed takeRecords as part of https://github.com/WICG/reporting/issues/58.
Filed https://github.com/WICG/reporting/issues/59 to discuss boot-time.